archive/100-theorems-list/16_abel_ruffini.lean

```diff
@@ -0,0 +1,179 @@
+  apply irreducible_of_eisenstein_criterion,
+  { rwa [ideal.span_singleton_prime (int.coe_nat_ne_zero.mpr hp.ne_zero),
+      int.prime_iff_nat_abs_prime] },
+  { rw [leading_coeff_Phi, ideal.mem_span_singleton],
+    exact_mod_cast mt nat.dvd_one.mp (hp.ne_one) },
+  { intros n hn,
+    rw ideal.mem_span_singleton,
+    rw [degree_Phi, with_bot.coe_lt_coe] at hn,
+    interval_cases n with hn;
+    simp [Φ, coeff_X_pow, coeff_C, int.coe_nat_dvd.mpr, hpb, hpa, -ring_hom.eq_int_cast] },
+  { simp only [degree_Phi, ←with_bot.coe_zero, with_bot.coe_lt_coe, nat.succ_pos'] },
+  { rwa [coeff_zero_Phi, pow_two, ideal.span_singleton_mul_span_singleton, ←pow_two,
+      ideal.mem_span_singleton, ←int.coe_nat_pow],
+    norm_num,
+    exact mt int.coe_nat_dvd.mp hp2b },
+  all_goals { exact polynomial.monic.is_primitive (monic_Phi a b) },
+end
+
+lemma real_roots_Phi_le : fintype.card ((Φ ℚ a b).root_set ℝ) ≤ 3 :=
+begin
+  rw [←map_Phi a b (algebra_map ℤ ℚ), Φ, ←one_mul (X ^ 5), ←C_1],
+  refine (card_root_set_le_derivative _).trans
+    (nat.succ_le_succ ((card_root_set_le_derivative _).trans (nat.succ_le_succ _))),
+  suffices : ((C ((algebra_map ℤ ℚ) 20) * X ^ 3).root_set ℝ).subsingleton,
+  { norm_num [fintype.card_le_one_iff_subsingleton, ← mul_assoc, *] at * },
+  rw root_set_C_mul_X_pow; norm_num,
+end
+
+lemma real_roots_Phi_ge_aux (hab : b < a) :
+  ∃ x y : ℝ, x ≠ y ∧ aeval x (Φ ℚ a b) = 0 ∧ aeval y (Φ ℚ a b) = 0 :=
+begin
+  let f := λ x : ℝ, aeval x (Φ ℚ a b),
+  have hf : f = λ x, x ^ 5 - a * x + b := by simp [f, Φ],
+  have hc : ∀ s : set ℝ, continuous_on f s := λ s, (Φ ℚ a b).continuous_on_aeval,
+  have ha : (1 : ℝ) ≤ a := nat.one_le_cast.mpr (nat.one_le_of_lt hab),
+  have hle : (0 : ℝ) ≤ 1 := zero_le_one,
+  have hf0 : 0 ≤ f 0 := by norm_num [hf],
+  by_cases hb : (1 : ℝ) - a + b < 0,
+  { have hf1 : f 1 < 0 := by norm_num [hf, hb],
+    have hfa : 0 ≤ f a,
+    { simp_rw [hf, ←sq],
+      refine add_nonneg (sub_nonneg.mpr (pow_le_pow ha _)) _; norm_num },
+    obtain ⟨x, ⟨-, hx1⟩, hx2⟩ := intermediate_value_Ico' hle (hc _) (set.mem_Ioc.mpr ⟨hf1, hf0⟩),
+    obtain ⟨y, ⟨hy1, -⟩, hy2⟩ := intermediate_value_Ioc ha (hc _) (set.mem_Ioc.mpr ⟨hf1, hfa⟩),
+    exact ⟨x, y, (hx1.trans hy1).ne, hx2, hy2⟩ },
+  { replace hb : (b : ℝ) = a - 1 := by linarith [show (b : ℝ) + 1 ≤ a, by exact_mod_cast hab],
+    have hf1 : f 1 = 0 := by norm_num [hf, hb],
+    have hfa := calc f (-a) = (-a) ^ 5 - a * -a + (a - 1) : by rw [hf, hb]
+      ... ≤ (-a) ^ 3 - a * -a + (a - 1) : _
+      ... = -((a - 1) ^ 2 * (a + 1)) : by ring
+      ... ≤ 0 : by nlinarith,
+    { obtain ⟨x, ⟨-, hx1⟩, hx2⟩ := intermediate_value_Icc
+        (neg_nonpos.mpr (hle.trans ha)) (hc _) (set.mem_Icc.mpr ⟨hfa, hf0⟩),
+      exact ⟨x, 1, (hx1.trans_lt zero_lt_one).ne, hx2, hf1⟩ },
+    { rw [add_le_add_iff_right, sub_le_sub_iff_right, neg_pow_bit1, neg_pow_bit1],
+      exact neg_le_neg (pow_le_pow ha (bit1_le_bit1.mpr one_le_two)) } },
```

Comment:
Normally I wouldn't be so picky, but since this will likely be a flagship theorem in the archive can you please make the following change to improve style and readability? (It's actually one line shorter as well)
```suggestion
    have hfa := calc f (-a) = a ^ 2 - a ^ 5 + b       : by norm_num [hf, ← sq]
                        ... ≤ a ^ 2 - a ^ 3 + (a - 1) : by refine add_le_add (sub_le_sub_left
                                                            (pow_le_pow ha _) _) _; linarith
                        ... = -(a - 1) ^ 2 * (a + 1)  : by ring
                        ... ≤ 0                       : by nlinarith,
    have ha' := neg_nonpos.mpr (hle.trans ha),
    obtain ⟨x, ⟨-, hx1⟩, hx2⟩ := intermediate_value_Icc ha' (hc _) (set.mem_Icc.mpr ⟨hfa, hf0⟩),
    exact ⟨x, 1, (hx1.trans_lt zero_lt_one).ne, hx2, hf1⟩ },
```